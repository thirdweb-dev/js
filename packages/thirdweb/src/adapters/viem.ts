import type { Abi } from "abitype";
import {
  http,
  type GetContractReturnType,
  type PublicClient,
  type TransactionSerializableEIP1559,
  type Chain as ViemChain,
  type WalletClient,
  createPublicClient,
  createWalletClient,
  custom,
} from "viem";
import type { Chain } from "../chains/types.js";
import { getRpcUrlForChain } from "../chains/utils.js";
import type { ThirdwebClient } from "../client/client.js";
import { resolveContractAbi } from "../contract/actions/resolve-abi.js";
import { type ThirdwebContract, getContract } from "../contract/contract.js";
import { getRpcClient } from "../rpc/rpc.js";
import { estimateGas } from "../transaction/actions/estimate-gas.js";
import { sendTransaction } from "../transaction/actions/send-transaction.js";
import { prepareTransaction } from "../transaction/prepare-transaction.js";
import type { Account } from "../wallets/interfaces/wallet.js";

export const viemAdapter = {
  contract: {
    /**
     * Creates a ThirdwebContract from a Viem contract.
     * @param options - The options for creating the contract.
     * @returns The ThirdwebContract.
     * @example
     * ```ts
     * import { viemAdapter } from "thirdweb/adapters";
     *
     * const contract = viemmAdapter.contract.fromViem({
     *  viemContract: viemContract,
     *  chain: ethereum,
     *  client,
     * });
     * ```
     */
    fromViem: fromViemContract,
    /**
     * Converts a ThirdwebContract instance to a Viem contract representation.
     * @param contract The ThirdwebContract instance to convert.
     * @returns A promise that resolves to the Viem contract representation.
     * @example
     * ```ts
     * import { viemAdapter } from "thirdweb/adapters";
     *  const viemContract = await viemAdapter.contract.toViem({ thirdwebContract });
     * ```
     */
    toViem: toViemContract,
  },

  /**
   * Converts options to a Viem public client.
   * @param options - The options for creating the Viem public client.
   * @returns The Viem public client.
   * @example
   * ```ts
   * import { viemAdapter } from "thirdweb/adapters";
   *
   *  const publicClient = viemAdapter.publicClient.toViem({
   *  chain: ethereum,
   *  client,
   * });
   * ```
   */
  publicClient: {
    toViem: toViemPublicClient,
  },

  walletClient: {
    /**
     * Converts options to a Viem Wallet client.
     * @param options - The options for creating the Viem Wallet client.
     * @returns The Viem Wallet client.
     * @example
     * ```ts
     * import { viemAdapter } from "thirdweb/adapters";
     *
     * const walletClient = viemAdapter.walletClient.toViem({
     *  wallet: wallet,
     *  client,
     *  chain: ethereum,
     * });
     * ```
     */
    toViem: toViemWalletClient,

    /**
     * Converts a Viem Wallet client to an Account.
     * @param options - The options for creating the Account.
     * @returns The Account.
     * @example
     * ```ts
     * import { viemAdapter } from "thirdweb/adapters";
     *
     * const account = viemAdapter.walletClient.fromViem({
     *   walletClient,
     * });
     */
    fromViem: fromViemWalletClient,
  },
};

type FromViemContractOptions<TAbi extends Abi> = {
  client: ThirdwebClient;
  viemContract: GetContractReturnType<TAbi>;
  chain: Chain;
};

function fromViemContract<const TAbi extends Abi>(
  options: FromViemContractOptions<TAbi>,
): ThirdwebContract<TAbi> {
  return getContract({
    address: options.viemContract.address,
    abi: options.viemContract.abi,
    chain: options.chain,
    client: options.client,
  });
}

async function toViemContract<const TAbi extends Abi>(options: {
  thirdwebContract: ThirdwebContract<TAbi>;
}): Promise<GetContractReturnType<TAbi>> {
  return {
    address: options.thirdwebContract.address,
    abi: await resolveContractAbi(options.thirdwebContract),
  };
}

type ToViemPublicClientOptions = {
  client: ThirdwebClient;
  chain: Chain;
};

function toViemPublicClient(options: ToViemPublicClientOptions): PublicClient {
  const { chain, client } = options;
  const rpcUrl = getRpcUrlForChain({ chain, client });
  const viemChain: ViemChain = {
    id: chain.id,
    name: chain.name || "",
    rpcUrls: {
      default: { http: [rpcUrl] },
    },
    nativeCurrency: {
      name: chain.nativeCurrency?.name || "Ether",
      symbol: chain.nativeCurrency?.symbol || "ETH",
      decimals: chain.nativeCurrency?.decimals || 18,
    },
  };
  return createPublicClient({
    transport: http(rpcUrl, {
      fetchOptions: client.secretKey
        ? {
            headers: {
              "x-secret-key": client.secretKey,
            },
          }
        : undefined,
    }),
    chain: viemChain,
  });
}

type ToViemWalletClientOptions = {
  account: Account;
  client: ThirdwebClient;
  chain: Chain;
};

function toViemWalletClient(options: ToViemWalletClientOptions): WalletClient {
  const { account, chain, client } = options;
  if (!account) {
    throw new Error("Wallet not connected.");
  }

  const rpcUrl = getRpcUrlForChain({ chain, client });
  const viemChain: ViemChain = {
    id: chain.id,
    name: chain.name || "",
    rpcUrls: {
      default: { http: [rpcUrl] },
    },
    nativeCurrency: {
      name: chain.nativeCurrency?.name || "Ether",
      symbol: chain.nativeCurrency?.symbol || "ETH",
      decimals: chain.nativeCurrency?.decimals || 18,
    },
  };

  const rpcClient = getRpcClient({ client, chain });
  const transport = custom({
    request: async (request) => {
      if (request.method === "eth_sendTransaction") {
        const result = await sendTransaction({
          transaction: prepareTransaction({
            ...request.params[0],
            chain,
            client,
          }),
          account: account,
        });
        return result.transactionHash;
      }
      if (request.method === "eth_estimateGas") {
        return estimateGas({
          transaction: prepareTransaction({
            ...request.params[0],
            chain,
            client,
          }),
          account: account,
        });
      }
      if (request.method === "personal_sign") {
        return account.signMessage({
          message: {
            raw: request.params[0],
          },
        });
      }
      if (request.method === "eth_signTypedData_v4") {
        const data = JSON.parse(request.params[1]);
        return account.signTypedData(data);
      }
      if (request.method === "eth_accounts") {
        return [account.address];
      }
      return rpcClient(request);
    },
  });

  return createWalletClient({
    transport,
    account: account.address,
    chain: viemChain,
    key: "thirdweb-wallet",
  });
}

function fromViemWalletClient(options: {
  walletClient: WalletClient;
}): Account {
  const viemAccount = options.walletClient.account;
  if (!viemAccount) {
    throw new Error(
      "Account not found in walletClient, please pass it explicitely.",
    );
  }
  return {
    address: viemAccount.address,
    signMessage: async (msg) => {
      return options.walletClient.signMessage({
        account: viemAccount,
        ...msg,
      });
    },
    sendTransaction: async (tx) => {
      const tx1559 = tx as TransactionSerializableEIP1559; // TODO check other txTypes
      const txHash = await options.walletClient.sendTransaction({
        account: viemAccount,
        chain: options.walletClient.chain,
        ...tx1559,
      });
      return {
        transaction: tx,
        transactionHash: txHash,
      };
    },
    signTypedData(_typedData) {
      if (!_typedData) {
        throw new Error("Typed data is required to signTypedData");
      }
      // biome-ignore lint/suspicious/noExplicitAny: TODO: fix later
      return options.walletClient.signTypedData(_typedData as any);
    },
  };
}
