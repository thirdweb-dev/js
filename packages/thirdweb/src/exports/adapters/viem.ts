export { viemAdapter } from "../../adapters/viem.js";
