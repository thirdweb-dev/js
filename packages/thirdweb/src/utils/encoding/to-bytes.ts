import { cachedTextEncoder } from "../text-encoder.js";
import { assertSize } from "./helpers/assert-size.js";
import { charCodeToBase16 } from "./helpers/charcode-to-base-16.js";
import { type Hex, isHex } from "./helpers/is-hex.js";
import { type NumberToHexOpts, numberToHex, padHex } from "./hex.js";

type PadOptions = {
  dir?: "left" | "right";
  size?: number | null;
};

function padBytes(bytes: Uint8Array, { dir, size = 32 }: PadOptions = {}) {
  if (size === null) {
    return bytes;
  }
  if (bytes.length > size) {
    throw new Error(`Size overflow: ${bytes.length} > ${size}`);
  }
  const paddedBytes = new Uint8Array(size);
  for (let i = 0; i < size; i++) {
    const padEnd = dir === "right";
    paddedBytes[padEnd ? i : size - i - 1] =
      // biome-ignore lint/style/noNonNullAssertion: we know its there
      bytes[padEnd ? i : bytes.length - i - 1]!;
  }
  return paddedBytes;
}

export type ToBytesParameters = {
  /** Size of the output bytes. */
  size?: number;
};

/**
 * Converts a value to an array of bytes.
 * @param value - The value to convert.
 * @param opts - Optional parameters for the conversion.
 * @returns The array of bytes representing the value.
 * @example
 * ```ts
 * import { toBytes } from "thirdweb/utils";
 * const bytes = toBytes("0x1a4");
 * console.log(bytes); // Uint8Array(2) [ 1, 164 ]
 * ```
 * @utils
 */
export function toBytes(
  value: string | bigint | number | boolean | Hex,
  opts: ToBytesParameters = {},
): Uint8Array {
  switch (typeof value) {
    case "number":
    case "bigint":
      return numberToBytes(value, opts);
    case "boolean":
      return boolToBytes(value, opts);
    default:
      if (isHex(value)) {
        return hexToBytes(value, opts);
      }
      return stringToBytes(value, opts);
  }
}

export type BoolToBytesOpts = {
  /** Size of the output bytes. */
  size?: number;
};

/**
 * Converts a boolean value to a Uint8Array of bytes.
 * @param value - The boolean value to convert.
 * @param opts - Optional parameters for the conversion.
 * @returns The Uint8Array of bytes representing the boolean value.
 * @example
 * ```ts
 * import { boolToBytes } from "thirdweb/utils";
 * const bytes = boolToBytes(true);
 * console.log(bytes); // Uint8Array(1) [ 1 ]
 * ```
 * @utils
 */
export function boolToBytes(value: boolean, opts: BoolToBytesOpts = {}) {
  const bytes = new Uint8Array(1);
  bytes[0] = Number(value);
  if (typeof opts.size === "number") {
    assertSize(bytes, { size: opts.size });
    return padBytes(bytes, { size: opts.size });
  }
  return bytes;
}

export type HexToBytesOpts = {
  /** Size of the output bytes. */
  size?: number;
};

/**
 * Converts a hexadecimal string to a Uint8Array of bytes.
 * @param hex_ The hexadecimal string to convert.
 * @param opts Options for converting the hexadecimal string.
 * @returns The Uint8Array of bytes.
 * @throws Error if the byte sequence is invalid.
 * @example
 * ```ts
 * import { hexToBytes } from "thirdweb/utils";
 * const bytes = hexToBytes("0x1a4");
 * console.log(bytes); // Uint8Array(2) [ 1, 164 ]
 * ```
 * @utils
 */
export function hexToBytes(hex_: Hex, opts: HexToBytesOpts = {}): Uint8Array {
  let hex = hex_;
  if (opts.size) {
    assertSize(hex, { size: opts.size });
    hex = padHex(hex, { dir: "right", size: opts.size });
  }

  let hexString = hex.slice(2) as string;
  if (hexString.length % 2) {
    hexString = `0${hexString}`;
  }

  const length = hexString.length / 2;
  const bytes = new Uint8Array(length);
  for (let index = 0, j = 0; index < length; index++) {
    const nibbleLeft = charCodeToBase16(hexString.charCodeAt(j++));
    const nibbleRight = charCodeToBase16(hexString.charCodeAt(j++));
    if (nibbleLeft === undefined || nibbleRight === undefined) {
      throw new Error(
        `Invalid byte sequence ("${hexString[j - 2]}${
          hexString[j - 1]
        }" in "${hexString}").`,
      );
    }
    bytes[index] = nibbleLeft * 16 + nibbleRight;
  }
  return bytes;
}

/**
 * Converts a number to bytes.
 * @param value - The number to convert.
 * @param opts - Options for converting the number to hex.
 * @returns The bytes representation of the number.
 * @example
 * ```ts
 * import { numberToBytes } from "thirdweb/utils";
 * const bytes = numberToBytes(420);
 * console.log(bytes); // Uint8Array(2) [ 1, 164 ]
 * ```
 * @utils
 */
export function numberToBytes(value: bigint | number, opts?: NumberToHexOpts) {
  const hex = numberToHex(value, opts);
  return hexToBytes(hex);
}

export type StringToBytesOpts = {
  /** Size of the output bytes. */
  size?: number;
};

/**
 * Converts a string to an array of bytes.
 * @param value - The string to convert.
 * @param opts - Optional parameters for the conversion.
 * @returns The array of bytes representing the string.
 * @example
 * ```ts
 * import { stringToBytes } from "thirdweb/utils";
 * const bytes = stringToBytes("Hello, world!");
 * console.log(bytes); // Uint8Array(13) [ 72, 101, 108, 108, 111, 44, 32, 119, 111, 114, 108, 100, 33 ]
 * ```
 * @utils
 */
export function stringToBytes(
  value: string,
  opts: StringToBytesOpts = {},
): Uint8Array {
  const bytes = cachedTextEncoder().encode(value);
  if (typeof opts.size === "number") {
    assertSize(bytes, { size: opts.size });
    return padBytes(bytes, { dir: "right", size: opts.size });
  }
  return bytes;
}
