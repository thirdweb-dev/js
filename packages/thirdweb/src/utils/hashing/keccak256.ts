import { keccak_256 } from "@noble/hashes/sha3";
import {
  type Hex,
  hexToUint8Array,
  isHex,
  uint8ArrayToHex,
} from "../encoding/hex.js";

type To = "hex" | "bytes";

export type Keccak256Hash<TTo extends To> =
  | (TTo extends "bytes" ? Uint8Array : never)
  | (TTo extends "hex" ? Hex : never);

/**
 * Calculates the Keccak-256 hash of the given value.
 * @param value - The value to hash, either as a hexadecimal string or a Uint8Array.
 * @param to - The desired output format of the hash (optional). Defaults to 'hex'.
 * @returns The Keccak-256 hash of the value in the specified format.
 * @example
 * ```ts
 * import { keccak256 } from "thirdweb/utils";
 * const hash = keccak256("0x1234");
 * ```
 * @utils
 */
export function keccak256<TTo extends To = "hex">(
  value: Hex | Uint8Array,
  to?: TTo,
): Keccak256Hash<TTo> {
  const bytes = keccak_256(
    isHex(value, { strict: false }) ? hexToUint8Array(value) : value,
  );
  if (to === "bytes") {
    return bytes as Keccak256Hash<TTo>;
  }
  // default fall through to hex
  return uint8ArrayToHex(bytes) as Keccak256Hash<TTo>;
}
