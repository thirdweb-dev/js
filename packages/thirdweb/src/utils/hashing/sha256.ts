import { sha256 as noble_sha256 } from "@noble/hashes/sha256";
import {
  type Hex,
  hexToUint8Array,
  isHex,
  uint8ArrayToHex,
} from "../encoding/hex.js";

type To = "hex" | "bytes";

export type Sha256Hash<TTo extends To> =
  | (TTo extends "bytes" ? Uint8Array : never)
  | (TTo extends "hex" ? Hex : never);

/**
 * Calculates the SHA256 hash of the given value.
 * @param value - The value to hash. It can be either a hexadecimal string or a Uint8Array.
 * @param to - (Optional) The desired output format of the hash. Defaults to 'hex'.
 * @returns The SHA256 hash of the value in the specified format.
 * @example
 * ```ts
 * import { sha256 } from "thirdweb/utils";
 * const hash = sha256("0x1234");
 * ```
 * @utils
 */
export function sha256<TTo extends To = "hex">(
  value: Hex | Uint8Array,
  to?: TTo,
): Sha256Hash<TTo> {
  const bytes = noble_sha256(
    isHex(value, { strict: false }) ? hexToUint8Array(value) : value,
  );
  if (to === "bytes") {
    return bytes as Sha256Hash<TTo>;
  }
  return uint8ArrayToHex(bytes) as Sha256Hash<TTo>;
}
