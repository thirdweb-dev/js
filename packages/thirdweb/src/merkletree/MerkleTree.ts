// ADAPTED FROM https://github.com/merkletreejs/merkletreejs
import { type Hex, uint8ArrayToHex } from "../utils/encoding/hex.js";
import { hexToBytes } from "../utils/encoding/to-bytes.js";
import { keccak256 } from "../utils/hashing/keccak256.js";
import {
  areUint8ArraysEqual,
  compareUint8Arrays,
  concatUint8Arrays,
} from "../utils/uint8-array.js";

// TODO: clean this up more, there is probably something to be said to move this entirely out of a class

type TLeaf = Uint8Array;
type TLayer = Uint8Array[];

/**
 * Class reprensenting a Merkle Tree
 * @namespace MerkleTree
 */
export class MerkleTree {
  private leaves: TLeaf[] = [];
  private layers: TLayer[] = [];

  constructor(leaves: (Uint8Array | Hex)[]) {
    this.leaves = leaves.map((el) =>
      el instanceof Uint8Array ? el : hexToBytes(el),
    );

    this.leaves = this.leaves.sort();

    this.createHashes(this.leaves);
  }

  public getHexRoot(): `0x${string}` {
    return uint8ArrayToHex(this.getRoot());
  }

  public getHexProof(leaf: Uint8Array | Hex, index?: number): `0x${string}`[] {
    return this.getProof(leaf, index).map((item) => uint8ArrayToHex(item.data));
  }

  // private below

  private createHashes(nodes: Uint8Array[]) {
    this.layers = [nodes];
    while (nodes.length > 1) {
      const layerIndex = this.layers.length;

      this.layers.push([]);

      const layerLimit = nodes.length;

      for (let i = 0; i < nodes.length; i += 2) {
        if (i >= layerLimit) {
          // biome-ignore lint/style/noNonNullAssertion: in bounds
          this.layers[layerIndex]!.push(...nodes.slice(layerLimit));
          break;
        }
        if (i + 1 === nodes.length) {
          if (nodes.length % 2 === 1) {
            // push copy of hash and continue iteration
            // biome-ignore lint/style/noNonNullAssertion: in bounds
            this.layers[layerIndex]!.push(nodes[i]!);
            continue;
          }
        }

        // biome-ignore lint/style/noNonNullAssertion: in bounds
        const left = nodes[i]!;
        // biome-ignore lint/style/noNonNullAssertion: in bounds
        const right = i + 1 === nodes.length ? left : nodes[i + 1]!;
        const combined = [left, right];

        combined.sort(compareUint8Arrays);

        const hash = keccak256(concatUint8Arrays(combined), "bytes");

        // biome-ignore lint/style/noNonNullAssertion: in bounds
        this.layers[layerIndex]!.push(hash);
      }
      // biome-ignore lint/style/noParameterAssign: part of the functionality
      // biome-ignore lint/style/noNonNullAssertion: in bounds
      nodes = this.layers[layerIndex]!;
    }
  }

  private getRoot(): Uint8Array {
    if (this.layers.length === 0) {
      return new Uint8Array();
    }

    // biome-ignore lint/style/noNonNullAssertion: in bounds
    return this.layers[this.layers.length - 1]![0] || new Uint8Array();
  }

  private getProof(
    leaf: Uint8Array | Hex,
    index?: number,
  ): { position: "left" | "right"; data: Uint8Array }[] {
    if (typeof leaf === "undefined") {
      throw new Error("leaf is required");
    }
    // biome-ignore lint/style/noParameterAssign: part of the functionality
    leaf = leaf instanceof Uint8Array ? leaf : hexToBytes(leaf);
    const proof = [];

    if (!Number.isInteger(index)) {
      // biome-ignore lint/style/noParameterAssign: part of the functionality
      index = -1;

      for (let i = 0; i < this.leaves.length; i++) {
        // biome-ignore lint/style/noNonNullAssertion: within bounds
        if (areUint8ArraysEqual(leaf, this.leaves[i]!)) {
          // biome-ignore lint/style/noParameterAssign: part of the functionality
          index = i;
        }
      }
    }

    // @ts-expect-error - issue from original code
    if (index <= -1) {
      return [];
    }

    for (let i = 0; i < this.layers.length; i++) {
      // biome-ignore lint/style/noNonNullAssertion: in bounds
      const layer = this.layers[i]!;
      // @ts-expect-error - issue from original code
      const isRightNode = index % 2;
      const pairIndex = isRightNode
        ? // @ts-expect-error - issue from original code
          index - 1
        : // Proof Generation for Non-Bitcoin Trees
          // @ts-expect-error - issue from original code
          index + 1;

      if (pairIndex < layer.length) {
        proof.push({
          position: isRightNode ? "left" : "right",
          data: layer[pairIndex],
        });
      }

      // set index to parent index
      // @ts-expect-error - issue from original code
      // biome-ignore lint/style/noParameterAssign: part of the functionality
      index = (index / 2) | 0;
    }

    // @ts-expect-error - issue from original code
    return proof;
  }
}
