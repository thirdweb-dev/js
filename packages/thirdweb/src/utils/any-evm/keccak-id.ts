import type { Hex } from "../encoding/hex.js";
import { stringToBytes } from "../encoding/to-bytes.js";
import { keccak256 } from "../hashing/keccak256.js";

/**
 * Calculates the keccak ID of the given input.
 * @param input - The input value to calculate the keccak ID for.
 * @returns The keccak ID as a Hex string.
 * @example
 * ```ts
 * import { keccackId } from "thirdweb/utils";
 * const keccakId = keccackId(input);
 * ```
 * @utils
 */
export function keccakId(input: string): Hex {
  return keccak256(stringToBytes(input));
}
