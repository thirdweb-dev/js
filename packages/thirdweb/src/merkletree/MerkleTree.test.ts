import { concat } from "viem";
import { describe, expect, it as test } from "vitest";
import { keccakId } from "../utils/any-evm/keccak-id.js";
import { hexToUint8Array, uint8ArrayToHex } from "../utils/encoding/hex.js";
import { stringToBytes } from "../utils/encoding/to-bytes.js";
import { keccak256 } from "../utils/hashing/keccak256.js";
import { sha256 } from "../utils/hashing/sha256.js";
import { concatUint8Arrays } from "../utils/uint8-array.js";
import { MerkleTree } from "./MerkleTree.js";

describe("MerkleTree", () => {
  test("sha256 with keccak256 leaves", () => {
    const leaves = ["a", "b", "c"].map(keccakId);
    const tree = new MerkleTree(leaves, sha256);

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
    // check equality of the leaves
    expect(tree.getHexLeaves()).toMatchSnapshot();
    // check equality of the layers
    expect(tree.getHexLayers()).toMatchSnapshot();
  });

  test("sha256 with keccak256 leaves with duplicate odd option", () => {
    const leaves = ["a", "b", "c"].map(keccakId);
    const tree = new MerkleTree(leaves, sha256, { duplicateOdd: true });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
    // check equality of the leaves
    expect(tree.getHexLeaves()).toMatchSnapshot();
    // check equality of the layers
    expect(tree.getHexLayers()).toMatchSnapshot();
  });

  test("sha256 with sort pairs option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map((v) =>
      sha256(stringToBytes(v)),
    );
    const tree = new MerkleTree(leaves, sha256, { sortPairs: true });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  test("sha256 verify with positional hex proof and no pairSort", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map((v) =>
      sha256(stringToBytes(v)),
    );
    const tree = new MerkleTree(leaves, sha256, { sortPairs: false });

    expect(
      tree.verify(
        // biome-ignore lint/style/noNonNullAssertion: test
        tree.getProof(leaves[1]!, 1),
        // biome-ignore lint/style/noNonNullAssertion: test
        leaves[1]!,
        tree.getHexRoot(),
      ),
    ).toBe(true);
  });

  test("sha256 verify with non-hex proof and no pairSort", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map((v) =>
      sha256(stringToBytes(v)),
    );
    const tree = new MerkleTree(leaves, sha256, { sortPairs: false });

    expect(
      // biome-ignore lint/style/noNonNullAssertion: <explanation>
      tree.verify(tree.getProof(leaves[1]!, 1), leaves[1]!, tree.getHexRoot()),
    ).toBe(true);
  });

  test("sha256 verify with hex proof and pairSort", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map((v) =>
      sha256(stringToBytes(v)),
    );
    const tree = new MerkleTree(leaves, sha256, { sortPairs: true });

    expect(
      tree.verify(
        // TODO: accept Uint8Array as input
        tree.getHexProof(leaves[1]!, 1),
        // TODO: accept Uint8Array as input
        leaves[1]!,
        tree.getHexRoot(),
      ),
    ).toBe(true);
  });

  test("keccak256 with sort leaves and sort pairs option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"].map(
      keccakId,
    );
    const tree = new MerkleTree(leaves, keccakId, {
      sortLeaves: true,
      sortPairs: true,
    });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  test("keccak256 with sort option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j"].map(
      keccakId,
    );
    const tree = new MerkleTree(leaves, keccakId, { sort: true });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  test("sha256 with sha256 leaves and sort pairs option and duplicate odd option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"].map((v) =>
      sha256(stringToBytes(v)),
    );
    const tree = new MerkleTree(leaves, sha256, {
      sortPairs: true,
      duplicateOdd: true,
    });

    // check equality of the roots
    expect(tree.getHexRoot()).toMatchSnapshot();
  });

  test("sha256 with hash leaves option", () => {
    const expectedRoot =
      "0x1f7379539707bcaea00564168d1d4d626b09b73f8a2a365234c62d763f854da2";

    const leaves = ["a", "b", "c", "d", "e", "f"];
    const tree1 = new MerkleTree(leaves, sha256, { hashLeaves: true });
    expect(tree1.getHexRoot()).toEqual(expectedRoot);

    const hashedLeaves = ["a", "b", "c", "d", "e", "f"].map((v) =>
      sha256(stringToBytes(v)),
    );
    const tree2 = new MerkleTree(hashedLeaves, sha256);
    expect(tree2.getHexRoot()).toEqual(expectedRoot);
  });

  test("sha256 with hash leaves option and duplicate odd option", () => {
    const leaves = ["a", "b", "c", "d", "e", "f"];
    const tree = new MerkleTree(leaves, sha256, {
      hashLeaves: true,
      duplicateOdd: true,
    });
    const root =
      "0x44205acec5156114821f1f71d87c72e0de395633cd1589def6d4444cc79f8103";

    expect(tree.getHexRoot()).toEqual(root);
  });

  test("solidity keccak256", () => {
    const leaves = ["a", "b", "c"].map(keccakId);

    const a_hash =
      "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb";
    const b_hash =
      "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510";
    const c_hash =
      "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2";

    expect(leaves).toEqual([a_hash, b_hash, c_hash]);

    const tree = new MerkleTree(leaves, keccakId);

    const layers = tree.getHexLayers().slice(1); // no leaves

    const layer_1 = "0xa823e1adfad5920f542db402eb9598f9c0464220fe0293df1a007424061cd1df"

    expect(layers[0]![0]!).toEqual(layer_1);
    expect(layers[0]![1]!).toEqual(c_hash);

    const root = "0x3c0f3e5eaffaf5aa01e231c75b30139492d251862c4716b24d2bbb39deea4f6e";
    expect(tree.getHexRoot()).toEqual(root);

    const proof_0 = tree.getProof(leaves[0]!);
    expect(proof_0.length).toEqual(2);
    expect(proof_0[0]!.position).toEqual("right");
    expect(uint8ArrayToHex(proof_0[0]!.data)).toEqual(b_hash);
    expect(proof_0[1]!.position).toEqual("right");
    expect(uint8ArrayToHex(proof_0[1]!.data)).toEqual(c_hash);

    expect(tree.verify(proof_0, leaves[0]!, root)).toBe(true);

    const proof_1 = tree.getProof(leaves[1]!);
    expect(proof_1.length).toEqual(2);
    expect(proof_1[0]!.position).toEqual("left");
    expect(uint8ArrayToHex(proof_1[0]!.data)).toEqual(a_hash);
    expect(proof_1[1]!.position).toEqual("right");
    expect(uint8ArrayToHex(proof_1[1]!.data)).toEqual(c_hash);

    expect(tree.verify(proof_1, leaves[1]!, root)).toBe(true);

    const proof_2 = tree.getProof(leaves[2]!);
    expect(proof_2.length).toEqual(1);
    expect(proof_2[0]!.position).toEqual("left");
    expect(uint8ArrayToHex(proof_2[0]!.data)).toEqual(layer_1);

    expect(tree.verify(proof_2, leaves[2]!, root)).toBe(true);
  });
});
