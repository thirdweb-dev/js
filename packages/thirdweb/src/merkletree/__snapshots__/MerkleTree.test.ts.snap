// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MerkleTree > keccak256 with sort leaves and sort pairs option 1`] = `"0x9788fdf3a3365b8b0871ea3dd057e0eb3bbde4c15a7112bcae7c51030d6c0729"`;

exports[`MerkleTree > keccak256 with sort option 1`] = `"0x9788fdf3a3365b8b0871ea3dd057e0eb3bbde4c15a7112bcae7c51030d6c0729"`;

exports[`MerkleTree > sha256 with keccak256 leaves 1`] = `"0x311d2e46f49b15fff8b746b74ad57f2cc9e0d9939fda94387141a2d3fdf187ae"`;

exports[`MerkleTree > sha256 with keccak256 leaves 2`] = `
[
  "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
  "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
  "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
]
`;

exports[`MerkleTree > sha256 with keccak256 leaves 3`] = `
[
  [
    "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
    "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
    "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
  ],
  [
    "0x176f0f307632fdd5831875eb709e2f68d770b102262998b214ddeb3f04164ae1",
    "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
  ],
  [
    "0x311d2e46f49b15fff8b746b74ad57f2cc9e0d9939fda94387141a2d3fdf187ae",
  ],
]
`;

exports[`MerkleTree > sha256 with keccak256 leaves with duplicate odd option 1`] = `"0xbcdd0f60308db788712205115fe4273bfda49fa0925611fee765a63df9ab96a1"`;

exports[`MerkleTree > sha256 with keccak256 leaves with duplicate odd option 2`] = `
[
  "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
  "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
  "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
]
`;

exports[`MerkleTree > sha256 with keccak256 leaves with duplicate odd option 3`] = `
[
  [
    "0x3ac225168df54212a25c1c01fd35bebfea408fdac2e31ddd6f80a4bbf9a5f1cb",
    "0xb5553de315e0edf504d9150af82dafa5c4667fa618ed0a6f19c69b41166c5510",
    "0x0b42b6393c1f53060fe3ddbfcd7aadcca894465a5a438f69c87d790b2299b9b2",
  ],
  [
    "0x176f0f307632fdd5831875eb709e2f68d770b102262998b214ddeb3f04164ae1",
    "0x43e061172b1177f25d0f156b2d2ed11728006fade8e167ff3d1b9dbc979a3358",
  ],
  [
    "0xbcdd0f60308db788712205115fe4273bfda49fa0925611fee765a63df9ab96a1",
  ],
]
`;

exports[`MerkleTree > sha256 with sha256 leaves and sort pairs option and duplicate odd option 1`] = `"0xadcf243ef947fcbf3fbf692d7bce9f31f17f3626dc6cce5261991502cc5b2caa"`;

exports[`MerkleTree > sha256 with sort pairs option 1`] = `"0x28168ee78e90a723d85aa5e62b1ad2108e5b58eb1c9aa278189028a33c6647bb"`;
